File: dataset/CoT/GEMINI_gemini-2.5-pro/Task91.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <cstring>
#include <stdexcept>

#ifdef _WIN32
    #include <winsock2.h>
    #include <ws2tcpip.h>
    #pragma comment(lib, "ws2_32.lib")
#else
    #include <sys/socket.h>
    #include <netinet/in.h>
    #include <arpa/inet.h>
    #include <unistd.h>
    typedef int SOCKET;
    const int INVALID_SOCKET = -1;
    const int SOCKET_ERROR = -1;
    #define closesocket close
#endif

class Task91 {
private:
    static const int PORT = 5353;
    static const int BUFFER_SIZE = 512;
    static std::map<std::string, std::string> dnsRecords;

    static std::string parseDomainName(const unsigned char* buffer, int& pos, int buffer_len) {
        std::string name = "";
        int current_pos = pos;
        
        while (current_pos < buffer_len && buffer[current_pos] != 0) {
            unsigned char len = buffer[current_pos];
            if (len == 0) break;
            
            // Check for name compression
            if ((len & 0xC0) == 0xC0) {
                 if (current_pos + 1 >= buffer_len) throw std::runtime_error("Malformed packet: pointer out of bounds");
                int pointer = ((len & 0x3F) << 8) + buffer[current_pos + 1];
                int dummy_pos = pointer;
                name += parseDomainName(buffer, dummy_pos, buffer_len);
                pos = current_pos + 2; // Move original position past the 2-byte pointer
                return name;
            }

            if (current_pos + 1 + len > buffer_len) throw std::runtime_error("Malformed packet: label out of bounds");
            
            name.append((const char*)&buffer[current_pos + 1], len);
            name.append(".");
            current_pos += (len + 1);
        }
        
        pos = (buffer[current_pos] == 0) ? current_pos + 1 : current_pos;
        return name;
    }

    static std::vector<unsigned char> buildDnsResponse(const unsigned char* request, int requestLen) {
        if (requestLen < 12) {
            std::cerr << "Request too short." << std::endl;
            return {};
        }

        // --- Parse request ---
        int pos = 12;
        std::string domainName;
        try {
            domainName = parseDomainName(request, pos, requestLen);
        } catch (const std::runtime_error& e) {
            std::cerr << e.what() << std::endl;
            return {};
        }
        
        if (pos + 4 > requestLen) {
             std::cerr << "Request missing QTYPE/QCLASS." << std::endl;
             return {};
        }

        uint16_t qtype = (request[pos] << 8) | request[pos+1];
        uint16_t qclass = (request[pos+2] << 8) | request[pos+3];

        if (qtype != 1 || qclass != 1 || dnsRecords.find(domainName) == dnsRecords.end()) {
            std::cout << "Unsupported query or unknown domain: " << domainName << std::endl;
            return {};
        }
        std::cout << "Query for: " << domainName << std::endl;
        
        // --- Build response ---
        std::vector<unsigned char> response;
        response.insert(response.end(), request, request + 12); // Copy header

        // Set response flags (QR=1, Opcode=0, AA=1, RCODE=0 -> 0x8400)
        response[2] = 0x84;
        response[3] = 0x00;

        // Set ANCOUNT to 1
        response[6] = 0x00;
        response[7] = 0x01;

        // Copy question section
        int questionLen = (pos + 4) - 12;
        response.insert(response.end(), request + 12, request + 12 + questionLen);

        // Add Answer Section
        // Name pointer (to offset 12)
        response.push_back(0xC0);
        response.push_back(0x0C);
        
        // Type (A), Class (IN)
        response.push_back(0x00); response.push_back(0x01);
        response.push_back(0x00); response.push_back(0x01);

        // TTL (60s)
        uint32_t ttl = htonl(60);
        unsigned char* ttl_bytes = (unsigned char*)&ttl;
        response.insert(response.end(), ttl_bytes, ttl_bytes + 4);

        // RDLENGTH (4)
        response.push_back(0x00); response.push_back(0x04);
        
        // RDATA (IP address)
        in_addr addr;
        inet_pton(AF_INET, dnsRecords[domainName].c_str(), &addr);
        unsigned char* ip_bytes = (unsigned char*)&addr.s_addr;
        response.insert(response.end(), ip_bytes, ip_bytes + 4);

        return response;
    }

public:
    static void startDnsServer(const char* ip, int port) {
        #ifdef _WIN32
        WSADATA wsaData;
        if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
            std::cerr << "WSAStartup failed." << std::endl;
            return;
        }
        #endif

        SOCKET sock = socket(AF_INET, SOCK_DGRAM, 0);
        if (sock == INVALID_SOCKET) {
            std::cerr << "Socket creation failed." << std::endl;
            return;
        }

        sockaddr_in server_addr;
        server_addr.sin_family = AF_INET;
        server_addr.sin_port = htons(port);
        inet_pton(AF_INET, ip, &server_addr.sin_addr);

        if (bind(sock, (struct sockaddr*)&server_addr, sizeof(server_addr)) == SOCKET_ERROR) {
            std::cerr << "Bind failed." << std::endl;
            closesocket(sock);
            return;
        }

        std::cout << "DNS server listening on " << ip << ":" << port << std::endl;

        unsigned char buffer[BUFFER_SIZE];
        sockaddr_in client_addr;
        socklen_t client_len = sizeof(client_addr);

        while (true) {
            int recv_len = recvfrom(sock, (char*)buffer, BUFFER_SIZE, 0, (struct sockaddr*)&client_addr, &client_len);
            if (recv_len == SOCKET_ERROR) {
                std::cerr << "recvfrom failed." << std::endl;
                continue;
            }

            char client_ip[INET_ADDRSTRLEN];
            inet_ntop(AF_INET, &client_addr.sin_addr, client_ip, INET_ADDRSTRLEN);
            std::cout << "Received request from " << client_ip << std::endl;

            std::vector<unsigned char> response = buildDnsResponse(buffer, recv_len);
            if (!response.empty()) {
                sendto(sock, (const char*)response.data(), response.size(), 0, (struct sockaddr*)&client_addr, client_len);
            }
        }
        
        closesocket(sock);
        #ifdef _WIN32
        WSACleanup();
        #endif
    }
};

std::map<std::string, std::string> Task91::dnsRecords = {
    {"test1.local.", "192.168.1.1"},
    {"test2.local.", "192.168.1.2"},
    {"example.com.", "93.184.216.34"},
    {"my-server.net.", "10.0.0.1"},
    {"localhost.", "127.0.0.1"}
};

int main() {
    std::cout << "Starting C++ DNS Server..." << std::endl;
    std::cout << "You can test this server with commands like:" << std::endl;
    std::cout << "  dig @127.0.0.1 -p 5353 test1.local" << std::endl;
    std::cout << "  nslookup example.com 127.0.0.1" << std::endl;
    std::cout << "----------------------------------------------" << std::endl;
    Task91::startDnsServer("127.0.0.1", 5353);
    return 0;
}

File: dataset/CoT/GEMINI_gemini-2.5-pro/Task91.py
import socket
import struct

DNS_RECORDS = {
    b"test1.local.": "192.168.1.1",
    b"test2.local.": "192.168.1.2",
    b"example.com.": "93.184.216.34",
    b"my-server.net.": "10.0.0.1",
    b"localhost.": "127.0.0.1",
}
IP_ADDRESS = "127.0.0.1"
PORT = 5353
BUFFER_SIZE = 512

def parse_domain_name(data, offset):
    """Parses a domain name from a DNS query."""
    labels = []
    while True:
        length = data[offset]
        if length == 0:
            offset += 1
            break
        # Check for pointer compression
        if (length & 0xc0) == 0xc0:
            pointer = struct.unpack(">H", data[offset:offset+2])[0]
            pointer &= 0x3fff
            # Recursively parse the pointed-to name and stop current parsing
            labels.append(parse_domain_name(data, pointer)[0])
            offset += 2
            # Since a pointer terminates a name section, we break
            return b".".join(labels) + b'.', offset
        
        offset += 1
        labels.append(data[offset:offset+length])
        offset += length
    
    return b".".join(labels) + b'.', offset

def build_dns_response(data):
    """Parses the DNS query and builds a response."""
    # Transaction ID
    tid = data[:2]

    # Flags (Standard query)
    flags = data[2:4]
    
    # --- Parse Question ---
    # We assume 1 question
    try:
        domain_name, question_offset_end = parse_domain_name(data, 12)
        q_type, q_class = struct.unpack(">HH", data[question_offset_end:question_offset_end+4])
    except (IndexError, struct.error):
        print("Error parsing request: malformed packet.")
        return None

    # We only handle A record (type 1) queries for IN class (class 1)
    if q_type != 1 or q_class != 1:
        print(f"Unsupported query type ({q_type}) or class ({q_class})")
        return None

    ip_address_str = DNS_RECORDS.get(domain_name)
    print(f"Query for: {domain_name.decode('utf-8')}")

    if not ip_address_str:
        print(f"Domain not found: {domain_name.decode('utf-8')}")
        # In a real server, build an NXDOMAIN response
        return None
    
    # --- Build Response ---
    # Response flags (QR=1, Opcode=0, AA=1, RCODE=0 -> 0x8400)
    # 0x8180 for standard response with recursion available
    response_flags = b'\x84\x00'

    # Header: tid, flags, qdcount=1, ancount=1, nscount=0, arcount=0
    header = tid + response_flags + b'\x00\x01\x00\x01\x00\x00\x00\x00'
    
    # Question section (copied from request)
    question = data[12:question_offset_end+4]

    # Answer section
    # Name (pointer to question name at offset 12 -> 0xc00c)
    name_pointer = b'\xc0\x0c'
    # Type=A (1), Class=IN (1), TTL=60, RDLENGTH=4
    answer_meta = struct.pack(">HHILH", 1, 1, 60, 4)
    # RDATA (IP address)
    ip_parts = [int(p) for p in ip_address_str.split('.')]
    ip_data = struct.pack("BBBB", *ip_parts)

    answer = name_pointer + answer_meta + ip_data

    return header + question + answer

def start_dns_server(ip, port):
    """Starts the DNS server on the specified IP and port."""
    try:
        # AF_INET for IPv4, SOCK_DGRAM for UDP
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock.bind((ip, port))
        print(f"DNS server listening on {ip}:{port}")

        while True:
            try:
                data, addr = sock.recvfrom(BUFFER_SIZE)
                print(f"Received request from {addr[0]}")
                
                response = build_dns_response(data)
                if response:
                    sock.sendto(response, addr)
            except Exception as e:
                print(f"An error occurred while handling a request: {e}")

    except OSError as e:
        print(f"Error starting server: {e}")
    finally:
        if 'sock' in locals():
            sock.close()

if __name__ == '__main__':
    print("Starting Python DNS Server...")
    print("You can test this server with commands like:")
    print(f"  dig @{IP_ADDRESS} -p {PORT} test1.local")
    print(f"  nslookup example.com {IP_ADDRESS}")
    print("----------------------------------------------")
    start_dns_server(IP_ADDRESS, PORT)